db/schema.sql

```diff
@@ -193,13 +193,14 @@ CREATE INDEX collect_jobs_interval_containment_index ON collect_jobs USING gist
 CREATE TABLE aggregate_share_jobs(
     id                      BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY, -- artificial ID, internal-only
     task_id                 BIGINT NOT NULL,    -- the task ID being collected
-    batch_interval          TSRANGE NOT NULL,   -- the batch interval, as a range of timestamps
+    batch_identifier        BYTEA NOT NULL,     -- encoded query-type-specific batch identifier (corresponds to identifier in BatchSelector)
+    batch_interval          TSRANGE,            -- batch interval, as a TSRANGE, populated only for time-interval tasks. (will always match batch_identifier)
```

Comment:
If I understand correctly:
- `batch_interval` will either be `NULL` or it will be identical to the value encoded into `batch_identifier`
- the only reason we keep this redundant representation is so that we can use Postgres' range operators within queries, as otherwise we would have to decode a time range from `batch_identifier` and do time interval intersection checks in code

I'm concerned about the possibility of the two columns disagreeing and I wonder if we can make this safer. A couple ideas off the top of my head:
- have columns `batch_interval TSRANGE` and `batch_identifier BIGINT` and use a [CHECK constraint](https://www.postgresql.org/docs/14/ddl-constraints.html#DDL-CONSTRAINTS-CHECK-CONSTRAINTS) on the table to ensure exactly one of them is non-NULL
- Keep the schema as-is, but use some more complex constraint to verify that `batch_identifier` is the correct encoding of the current value of `batch_interval`, if `batch_interval != NULL`
- Use a `TRIGGER` so that on any update to `batch_identifier`, we automatically set `batch_interval` to the correct value (this may not be possible since Postgres doesn't understand the DAP encoding scheme, but maybe we can register a callback to our code?).

I have to admit I don't fully understand the performance implications of any of these approaches.

Reply:
CHECK constraints are only evaluated upon inserts and updates, so I think it should be OK from a performance point of view. One hurdle for implementing constraints is that the query type is not in the tables that have batch identifiers, and the identifier encodings are not self-describing. For now, though, we can distinguish them by length. (`Interval` encodes to 16 bytes, and `BatchId` encodes to 32 bytes) It looks like we'd have to compose our own big-endian integer decoding expressions with `get_byte(bytea, integer)`, which would be gnarly, but feasible.

Reply:
(one more thought: in the longer term, we might decide that time-interval query types are no longer required. that would be a longer-term strategy to resolve the need to keep both of these fields around.)